docs/ansible/coding-conventions.md

```diff
@@ -0,0 +1,669 @@
+You can also assign variables to them! For instance, atlanta has its own NTP servers, so when setting up ntp.conf, we should use them.
+Let’s set those now:
+
+```yaml
+---
+# file: group_vars/atlanta
+ntp: ntp-atlanta.example.com
+backup: backup-atlanta.example.com
+```
+
+Variables aren’t just for geographic information either! Maybe the webservers have some configuration that doesn’t make sense for the database servers:
+
+```yaml
+---
+# file: group_vars/webservers
+apacheMaxRequestsPerChild: 3000
+apacheMaxClients: 900
+```
+
+If we had any default values, or values that were universally true, we would put them in a file called group_vars/all:
+
+```yaml
+---
+# file: group_vars/all
+ntp: ntp-boston.example.com
+backup: backup-boston.example.com
+```
+
+We can define specific hardware variance in systems in a host_vars file, but avoid doing this unless you need to:
+
+```yaml
+---
+# file: host_vars/db-bos-1.example.com
+foo_agent_port: 86
+bar_agent_port: 99
+```
+
+Again, if we are using dynamic inventory sources, many dynamic groups are automatically created.
+So a tag like “class:webserver” would load in variables from the file “group_vars/ec2_tag_class_webserver” automatically.
+
+## Top Level Playbooks Are Separated By Role
+
+In site.yml, we import a playbook that defines our entire infrastructure. This is a very short example, because it’s just importing some other playbooks:
+
+```yaml
+---
+# file: site.yml
+- import_playbook: webservers.yml
+- import_playbook: dbservers.yml
+```
+
+In a file like webservers.yml (also at the top level), we map the configuration of the webservers group to the roles performed by the webservers group:
+
+```yaml
+---
+# file: webservers.yml
+- hosts: webservers
+  roles:
+    - common
+    - webtier
+```
+
+The idea here is that we can choose to configure our whole infrastructure by “running” site.yml or we could just choose to run a subset by running webservers.yml.
+This is analogous to the “–limit” parameter to ansible but a little more explicit:
+
+```yaml
+ansible-playbook site.yml --limit webservers
+ansible-playbook webservers.yml
+```
+
+## Always Mention The State
+
+The ‘state’ parameter is optional to a lot of modules.
+Whether ‘state=present’ or ‘state=absent’, it’s always best to leave that parameter in your playbooks to make it clear, especially as some modules support additional states.
+
+## Bundling Ansible Modules With Playbooks
+
+If a playbook has a ./library directory relative to its YAML file, this directory can be used to add ansible modules that will automatically be in the ansible module path.
+This is a great way to keep modules that go with a playbook together.
+This is shown in the directory structure example at the start of this section.
+
+## Whitespace and Comments
+
+Generous use of whitespace to break things up, and use of comments (which start with ‘#’), is encouraged.
+
+## Optimize Playbook Execution
```

Comment:
for all sections add at least good examples. add bad examples if good are not self-explanatory enough

add separate section with include / import usage guidelines (correct usage may simlify and speed the playbook run very much)

group similar items (like all related to gather facts has to be near each other)
use subheadings